File: composite-builds/temple-gradle-plugin/src/main/java/com/jamesfchen/T1plugin.java
package com.jamesfchen;

import com.android.build.api.transform.Transform;
import com.android.build.gradle.BaseExtension;
import com.android.build.gradle.internal.pipeline.TransformManager;
import com.kronos.plugin.base.BaseTransform;

import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Copyright ® $ 2017
 * All right reserved.
 *
 * @author: jamesfchen
 * @since: Apr/19/2022  Tue
 */
public class T1plugin implements Plugin<Project> {
    @Override
    public void apply(Project project) {
        project.getExtensions().findByType(BaseExtension.class)
                .registerTransform(new LTransform());
    }

    static class LTransform extends Transform {
        @Override
        public boolean isIncremental() {
            return true;
        }

        @Override
        public String getName() {
            return "IbcTransform";
        }

        @Override
        public java.util.Set<com.android.build.api.transform.QualifiedContent.ContentType> getInputTypes() {
            return TransformManager.CONTENT_CLASS;
        }

        @Override
        public java.util.Set<? super com.android.build.api.transform.QualifiedContent.Scope> getScopes() {
            return TransformManager.SCOPE_FULL_PROJECT;
        }

        @Override
        public void transform(com.android.build.api.transform.TransformInvocation invocation) {
            BaseTransform baseTransform = new BaseTransform(invocation, new com.kronos.plugin.base.TransformCallBack() {
                @Nullable
                @Override
                public byte[] process(@NotNull String className, @Nullable byte[] classBytes) {
                    return null;
                }
            }, false);
            baseTransform.setDeleteCallBack(new com.kronos.plugin.base.DeleteCallBack() {
                @Override
                public void delete(String className, byte[] bytes) {
                }
            });
            baseTransform.openSimpleScan();
            baseTransform.startTransform();
        }
    }
}

File: composite-builds/temple-gradle-plugin/src/main/kotlin/com/jamesfchen/T1plugin.kt
package com.jamesfchen

import com.android.build.api.transform.TransformInvocation
import com.android.build.gradle.BaseExtension
import javassist.ClassPool
import org.gradle.api.Project

/**
 * Copyright ® $ 2017
 * All right reserved.
 *
 * @author: jamesfchen
 * @since: Apr/19/2022  Tue
 */
class T1plugin : FastInsertCodePlugin() {
    val classPool = ClassPool()
    override fun getName() = "T1plugin"
    lateinit var project: Project
    override fun apply(project: Project) {
        super.apply(project)
        this.project = project
    }

    override fun isIncremental() = false
    override fun onInsertCodeBegin(transformInvocation: TransformInvocation?) {
        project.extensions.findByType(BaseExtension::class.java)?.bootClasspath?.forEach {
            classPool.appendClassPath(it.absolutePath)
        }
    }

    override fun onInsertCode(info: ClassInfo?): ByteArray {
        //        info?.mather1?.absolutePath.let {
//            classPool.insertClassPath(it)
//        }
        if (info == null) return byteArrayOf()
        if ("com.jamesfchen.myhome.B" == info.canonicalName
            || "com.jamesfchen.b.C" == info.canonicalName
        ) {
//            val makeClass = classPool.makeClassIfNew(info.classStream)
            val makeClass = classPool.makeClass(info.classStream)
//                val makeClass = classPool.getCtClass(info.canonicalName)
            println("cjf ${makeClass.name}  ${info?.mather.absolutePath}")
            try {
                makeClass.defrost()
                val declaredMethod = makeClass.getDeclaredMethod("onCreate", null)
                val insertCode = "android.util.Log.d(\"cjf\",\"1\");"
                declaredMethod.insertBefore(insertCode)
//                    makeClass.writeFile(info.mather2?.absolutePath)
                return makeClass.toBytecode()
            } catch (e: Exception) {
                println("1 could not create onCreate() in ${makeClass.name};${info.mather?.exists()}   ${info.mather?.absolutePath}\n$e")
            } finally {
                makeClass.detach()
            }
        }
        return byteArrayOf()
    }

    override fun onInsertCodeEnd() {

    }
}